### PR TITLE
[runtime] Fix runtime behavior when dealing with missing symlinks

### PR DESCRIPTION
**Pull request by kumpera**

### CODE REVIEW

File: mcs/class/corlib/Test/System.IO/FileTest.cs [hunk 1]
```diff
@@ -2680,5 +2682,59 @@ public void MoveTest ()
...
+
+
+		[DllImport ("libc", SetLastError=true)]
+		public static extern int symlink (string oldpath, string newpath);
+
+		[DllImport ("libc", SetLastError=true)]
+		public static extern int rename (string oldpath, string newpath);
+
+		static void DeleteBadSymLink (string path) {
+			string newpath = path + new Random ().Next ();
```
Comment (akoeplinger) [inline mcs/class/corlib/Test/System.IO/FileTest.cs]: the `Random` instance should be shared, creating two instances in rapid succession like with the calls in the unit test most likely causes it to use the same seed.

File: mcs/class/corlib/Test/System.IO/FileTest.cs [hunk 2]
```diff
@@ -2680,5 +2682,59 @@ public void MoveTest ()
...
+		}
+
+		[Test]
+		public void SymLinkLoop ()
+		{
+			if (!RunningOnUnix)
+				Assert.Ignore ("Symlink are hard on windows");
+
+			var name1 = "f0ew4fuweo";
+			var name2 = "jneruyhbv32ir";
```
Comment (akoeplinger) [inline mcs/class/corlib/Test/System.IO/FileTest.cs]: appending a random number would be good here too, we do run multiple builds and test runs on the same machine.

File: mcs/class/corlib/Test/System.IO/FileTest.cs [hunk 3]
```diff
@@ -2680,5 +2682,59 @@ public void MoveTest ()
...
+		public static extern int rename (string oldpath, string newpath);
+
+		static void DeleteBadSymLink (string path) {
+			string newpath = path + new Random ().Next ();
+			rename (path, newpath);
+			File.Delete (newpath);
+		}
+
+		[Test]
+		public void SymLinkLoop ()
```
Comment (akoeplinger) [inline mcs/class/corlib/Test/System.IO/FileTest.cs]: Does this test work on XI/XA? If not we'll need to exclude it there.

File: mcs/class/corlib/Test/System.IO/FileTest.cs [hunk 4]
```diff
@@ -2680,5 +2682,59 @@ public void MoveTest ()
...
+					}
+				} catch (UnauthorizedAccessException ex) {
+					Assert.AreEqual (0x80070005u, (uint)ex.HResult, "Ensure HRESULT is correct");
+				}
+
+
+				try {
+					File.Delete (path1);
+				} catch (UnauthorizedAccessException ex) {
+					Assert.AreEqual (0x80070005u, (uint)ex.HResult, "Ensure HRESULT is correct");
```
Comment (akoeplinger) [inline mcs/class/corlib/Test/System.IO/FileTest.cs]: style nit: we can use new features in nunit(lite) now which makes this more readable:

```csharp
var ex = Assert.Throws<UnauthorizedAccessException> (() => File.Delete (path1));
Assert.AreEqual (0x80070005u, (uint)ex.HResult, "Ensure HRESULT is correct");
```

File: mcs/class/corlib/Test/System.IO/FileTest.cs (hunk 2 above)
Comment (kumpera) [inline mcs/class/corlib/Test/System.IO/FileTest.cs]: Good point, I'll stop being lazy and use proper unique file naming.

File: mcs/class/corlib/Test/System.IO/FileTest.cs (hunk 1 above)
Comment (kumpera) [inline mcs/class/corlib/Test/System.IO/FileTest.cs]: That's not relevant here, the point is just adding a suffix that won't collide with existing files.
Unless tests are executed in parallel, it won't matter as we rename & delete.

File: mcs/class/corlib/Test/System.IO/FileTest.cs (hunk 3 above)
Comment (kumpera) [inline mcs/class/corlib/Test/System.IO/FileTest.cs]: It should, we use those 2 libc functions in mono.

File: mcs/class/corlib/Test/System.IO/FileTest.cs (hunk 4 above)
Comment (kumpera) [inline mcs/class/corlib/Test/System.IO/FileTest.cs]: I have the very opposite opinion on using lambda based APIs.

**Comment by kumpera** [comment]
The above change should address your review @akoeplinger

File: mcs/class/corlib/Test/System.IO/FileTest.cs [hunk 5]
```diff
@@ -2680,5 +2683,73 @@ public void MoveTest ()
...
+			DeleteBadSymLink (path2);
+			try {
+				symlink (path1, path2);
+				symlink (path2, path1);
+
+				Assert.IsFalse (File.Exists (path1), "File.Exists must return false for symlink loops");
+
+				try {
+					using (var f = File.Open (path1, FileMode.Open, FileAccess.Read)) {
+						Assert.Fail ("File.Exists must throw for symlink loops");
```
Comment (akoeplinger) [inline mcs/class/corlib/Test/System.IO/FileTest.cs]: message mentions File.Exists but it should be File.Open

File: mcs/class/corlib/Test/System.IO/FileTest.cs [hunk 6]
```diff
@@ -2680,5 +2683,73 @@ public void MoveTest ()
...
+					}
+				} catch (UnauthorizedAccessException ex) {
+					Assert.AreEqual (0x80070005u, (uint)ex.HResult, "Ensure HRESULT is correct");
+				}
+
+
+				try {
+					File.Delete (path1);
```
Comment (akoeplinger) [inline mcs/class/corlib/Test/System.IO/FileTest.cs]: this should have Assert.Fail() otherwise we won't notice if this stops throwing.

**Comment by kumpera** [comment]
It turns out the original behavior I found was due to me not understanding windows permissions correctly when cross-validating.

This PR now adjusts us to it plus addresses the last set of reviews remarks on the test suite.

**Comment by kumpera** [comment]
This Linux i386 failure is this: ```
1) WaitAll_ManyTasks (MonoTests.System.Threading.Tasks.TaskTests.WaitAll_ManyTasks)
     Expected: True
  But was:  False

  at MonoTests.System.Threading.Tasks.TaskTests.WaitAll_ManyTasks () [0x0004d] in /mnt/jenkins/workspace/test-mono-pull-request-i386/mcs/class/corlib/Test/System.Threading.Tasks/TaskTest.cs:311 ```

I believe it to be unrelated.

**Pull request merged**